vowpalwabbit/cb_explore_adf_synthcover.cc

```diff
@@ -0,0 +1,224 @@
+
+public:
+  cb_explore_adf_synthcover(float epsilon, float psi, size_t synthcoversize, std::shared_ptr<rand_state> random_state,
+      VW::version_struct model_file_version);
+  ~cb_explore_adf_synthcover();
+
+  // Should be called through cb_explore_adf_base for pre/post-processing
+  void predict(VW::LEARNER::multi_learner& base, multi_ex& examples) { predict_or_learn_impl<false>(base, examples); }
+  void learn(VW::LEARNER::multi_learner& base, multi_ex& examples) { predict_or_learn_impl<true>(base, examples); }
+  void save_load(io_buf& model_file, bool read, bool text);
+
+private:
+  template <bool is_learn>
+  void predict_or_learn_impl(VW::LEARNER::multi_learner& base, multi_ex& examples);
+};
+
+cb_explore_adf_synthcover::cb_explore_adf_synthcover(float epsilon, float psi, size_t synthcoversize,
+    std::shared_ptr<rand_state> random_state, VW::version_struct model_file_version)
+    : _epsilon(epsilon)
+    , _psi(psi)
+    , _synthcoversize(synthcoversize)
+    , _random_state(random_state)
+    , _model_file_version(model_file_version)
+    , _action_probs(v_init<ACTION_SCORE::action_score>())
+    , min_cost(0.0)
+    , max_cost(0.0)
+{
+}
+
+template <bool is_learn>
+void cb_explore_adf_synthcover::predict_or_learn_impl(VW::LEARNER::multi_learner& base, multi_ex& examples)
+{
+  VW::LEARNER::multiline_learn_or_predict<is_learn>(base, examples, examples[0]->ft_offset);
+
+  const auto it =
+      std::find_if(examples.begin(), examples.end(), [](example* item) { return !item->l.cb.costs.empty(); });
+
+  if (it != examples.end())
+  {
+    const CB::cb_class logged = (*it)->l.cb.costs[0];
+
+    min_cost = std::min(logged.cost, min_cost);
+    max_cost = std::max(logged.cost, max_cost);
+  }
+
+  ACTION_SCORE::action_scores& preds = examples[0]->pred.a_s;
+  uint32_t num_actions = (uint32_t)examples.size();
+  if (num_actions == 0)
+  {
+    preds.clear();
+    return;
+  }
+  else if (num_actions == 1)
+  {
+    preds[0].score = 1;
+    return;
+  }
+
+  for (size_t i = 0; i < num_actions; i++) { preds[i].score = std::min(max_cost, std::max(min_cost, preds[i].score)); }
+  std::make_heap(
+      preds.begin(), preds.end(), [](const ACTION_SCORE::action_score& a, const ACTION_SCORE::action_score& b) {
+        return ACTION_SCORE::score_comp(&a, &b) > 0;
+      });
+
+  _action_probs.clear();
+  for (uint32_t i = 0; i < num_actions; i++) _action_probs.push_back({i, 0.});
+
+  policyaction.clear();
+  for (uint32_t i = 0; i < _synthcoversize;)
+  {
+    std::pop_heap(
+        preds.begin(), preds.end(), [](const ACTION_SCORE::action_score& a, const ACTION_SCORE::action_score& b) {
+          return ACTION_SCORE::score_comp(&a, &b) > 0;
+        });
+    // NB: what STL calls pop_back(), v_array calls pop().  facepalm.
+    auto minpred = preds.pop();
+
+    auto secondminpred = preds[0];
+    for (; secondminpred.score >= minpred.score && i < _synthcoversize; i++)
+    {
+      policyaction.push_back(minpred.action);
```

Comment:
I don't understand what `policyaction` is used for. I can only see it being cleared or having the `minipred.action` pushed to it but I can't see it being used anywhere else.

Reply:
This is vestigial, I'll get rid of it.  I was analytically centering but playing the uniform policy performed similarly and removed a hyperparameter.